web/src/infra/settings/CustomizationBar.tsx

```diff
@@ -0,0 +1,161 @@
+    this.setState({ customizationState });
+  }
+
+  render() {
+    return (
+      <div className={css.BarWrapper}>
+        {this.renderCustomizationDialog()}
+        {this.renderQuickCustomization()}
+        {this.renderFullCustomizationButton()}
+      </div>
+    );
+  }
+
+  private renderQuickCustomization() {
+    const custom = this.state.customization;
+    if (!custom || !custom.renderQuick) {
+      return null;
+    }
+    const mode = this.props.info.mode;
+    return custom.renderQuick({
+      mode,
+      currentValue: (this.state.customizationState || {})[mode]?.quick,
+      onChange: this._changeCustomValue(CustomizationType.QUICK),
+    });
+  }
+
+  private hasFullCustomization() {
+    const mode = this.props.info.mode;
+    const custom = this.state.customization;
+    if (!custom || !custom.renderFull) {
+      return false;
+    }
+    const fullCustom = custom.renderFull({
+      mode: this.props.info.mode,
+      currentValue: this.state.customizationState[mode]?.quick,
+      onChange: () => {},
+    });
+    return fullCustom !== null;
+  }
+
+  private renderFullCustomizationButton() {
+    if (!this.hasFullCustomization()) {
+      return null;
+    }
+    const mode = this.props.info.mode;
+    return (
+      <IconButton
+        aria-label="Customize game"
+        onClick={this.toggleFullCustomizationDialog}
+        style={{ height: '36px', width: '36px', marginLeft: 'auto' }}
+      >
+        <SettingsIcon style={{ color: this.state.customizationState[mode]?.full ? '#3f51b5' : undefined }} />
+      </IconButton>
+    );
+  }
+
+  private renderCustomizationDialog() {
+    if (!this.state.showCustomizationDialog) {
+      return null;
+    }
+    return (
+      <AlertLayer onClickaway={this.toggleFullCustomizationDialog}>
+        <Card className={css.CustomizationCard}>
+          {this.renderCustomizationDialogHeader()}
+          <div className={css.DialogContentWrapper}>{this.renderCustomizationDialogContent()}</div>
+        </Card>
+      </AlertLayer>
+    );
+  }
+
+  private renderCustomizationDialogContent() {
+    const custom = this.state.customization;
+    const mode = this.props.info.mode;
+    return custom.renderFull({
+      mode: this.props.info.mode,
+      currentValue: this.state.customizationState[mode]?.full,
+      onChange: this._changeCustomValue(CustomizationType.FULL),
+    });
+  }
+
+  private renderCustomizationDialogHeader() {
+    return (
+      <div className={css.DialogTitleWrapper}>
+        <IconButton aria-label="close" onClick={this.toggleFullCustomizationDialog} className={css.DialogClose}>
+          <CloseIcon />
+        </IconButton>
+        <Typography variant="h6" component="span" className={css.DialogTitle}>
+          Customization
+        </Typography>
+      </div>
+    );
+  }
+
+  toggleFullCustomizationDialog = () => {
+    this.setState({ showCustomizationDialog: !this.state.showCustomizationDialog });
```

Comment:
Careful with this... `this.setState()` does not immediately mutate state.  It instead creates a pending state transition.  Updating the state like this may cause flaky and hard-to-debug issues.

When the value of the new state depends on the value of the current state, I recommend you enqueue an atomic update, like this:

```typescript
this.setState((prevState => {
  return {showCustomizationDialog: !prevState.showCustomizationDialog};
}));
```

Reply:
Done.